src/renderers/shared/fiber/ReactFiberCommitWork.js

```diff
@@ -54,22 +55,37 @@ module.exports = function<T, P, I, TI, PI, C, CX, PL>(
 
   // Capture errors so they don't interrupt unmounting.
   function safelyCallComponentWillUnmount(current, instance) {
-    try {
-      instance.componentWillUnmount();
-    } catch (error) {
-      captureError(current, error);
+    let unmountError = null;
+    if (__DEV__) {
+      unmountError = invokeGuardedCallback(null, instance.componentWillUnmount, instance);
+    } else {
+      try {
+        instance.componentWillUnmount();
+      } catch (e) {
+        unmountError = e;
+      }
+    }
+    if (unmountError) {
+      captureError(current, unmountError);
     }
   }
 
-  // Capture errors so they don't interrupt unmounting.
   function safelyDetachRef(current : Fiber) {
-    try {
-      const ref = current.ref;
-      if (ref !== null) {
-        ref(null);
+    const ref = current.ref;
+    if (ref !== null) {
+      let refError = null;
+      if (__DEV__) {
+        refError = invokeGuardedCallback(null, ref, null, null);
+      } else {
+        try {
+          ref(null);
+        } catch (e) {
+          refError = e;
+        }
+      }
+      if (refError !== null) {
+        captureError(current, refError);
```

Comment:
Can we move this into the try/catch and only do the `if (ref !== null)` for the `__DEV__` case?